src/lib/mina_lib/mina_lib.ml

```diff
@@ -869,45 +870,83 @@ let best_chain ?max_length t =
+      Work_selector.pending_work_statements ~snark_pool ~fee_opt t.work_selector
       |> List.length
     in
     Mina_metrics.(
       Gauge.set Snark_work.pending_snark_work (Int.to_float pending_work))
   in
-  let spec =
-    One_or_two.map work.spec.instances
-      ~f:Snark_work_lib.Work.Single.Spec.statement
-  in
-  let cb _ =
-    (* remove it from seen jobs after attempting to adding it to the pool to avoid this work being reassigned
-     * If the diff is accepted then remove it from the seen jobs.
-     * If not then the work should have already been in the pool with a lower fee or the statement isn't referenced anymore or any other error. In any case remove it from the seen jobs so that it can be picked up if needed *)
-    Work_selector.remove t.snark_job_state spec
-  in
-  ignore (Or_error.try_with (fun () -> update_metrics ()) : unit Or_error.t) ;
-  Network_pool.Snark_pool.(
-    Local_sink.push t.pipes.snark_local_sink
-      (Resource_pool.Diff.of_result work, cb))
-  |> Deferred.don't_wait_for
+  match
+    Work_partitioner.submit_partitioned_work ~result:work
+      ~partitioner:t.work_partitioner
+  with
+  | SpecUnmatched ->
+      `SpecUnmatched
+  | Removed ->
+      `Removed
+  | Processed None ->
+      `Ok
+  | Processed (Some { spec_with_proof; fee; prover }) ->
+      let proofs = spec_with_proof |> One_or_two.map ~f:Tuple2.get2 in
+      let fee_with_prover = Fee_with_prover.{ fee; prover } in
+      let stmts =
+        spec_with_proof
+        |> One_or_two.map ~f:(fun (spec, _) ->
+               Snark_work_lib.Selector.Single.Spec.Poly.statement spec )
+      in
+      [%log' info t.config.logger] "Partitioner combined work"
+        ~metadata:
+          [ ( "stmts"
+            , One_or_two.to_yojson Mina_state.Snarked_ledger_state.to_yojson
+                stmts )
+          ; ("fee_with_prover", Fee_with_prover.to_yojson fee_with_prover)
+          ; ("proofs", proofs |> One_or_two.to_yojson Ledger_proof.to_yojson)
+          ] ;
+      ignore (Or_error.try_with (fun () -> update_metrics ()) : unit Or_error.t) ;
+      Network_pool.Snark_pool.(
+        Local_sink.push t.pipes.snark_local_sink
+          ( Add_solved_work
+              ( stmts
+              , Network_pool.Priced_proof.
+                  { proof = proofs; fee = fee_with_prover } )
+          , Result.iter_error ~f:(fun err ->
+                [%log' info t.config.logger]
+                  "Failed to push completed work to local snark sink, \
+                   returning them to work selector"
+                  ~metadata:
+                    [ ( "stmts"
+                      , One_or_two.to_yojson
+                          Mina_state.Snarked_ledger_state.to_yojson stmts )
+                    ; ("error", `String (Error.to_string_hum err))
+                    ] ;
+                let _spec = One_or_two.map ~f:Tuple2.get1 spec_with_proof in
+                failwith "TODO: bring back work to work selector here" ) ))
```

Comment:
This TODO needs to be resolved, I guess.
What can be a cause of failure? Hard to judge about whether we need to put data back to selector w/o knowing that precisely.